blockchain/consensus/deleg_cns/README.md

```diff
@@ -0,0 +1,240 @@
+          },
+          "SectorID": 0,
+          "Deal": {
+            "PieceCID": {
+              "/": "baga6ea4seaqemcrx5m3n7udmoz7x53aae5wnapqsxkk5n5ysvled3hwnoqm7wda"
+            },
+            "PieceSize": 2048,
+            "VerifiedDeal": false,
+            "Client": "t3w3sv7uykya4vkzk77d2h7lfuldtaa5rr3nvzh6fkutu5lesdwefgnqm2u3obpyrhhzbupl42kubzqmvfivjq",
+            "Provider": "t01000",
+            "Label": "0",
+            "StartEpoch": 0,
+            "EndEpoch": 9001,
+            "StoragePricePerEpoch": "0",
+            "ProviderCollateral": "0",
+            "ClientCollateral": "0"
+          },
+          "DealClientKey": {
+            "Type": "bls",
+            "PrivateKey": "NOrAYwVSGwX2vtQ9MMySUV8YFxKiO97mGRALrejE/Tk=",
+            "PublicKey": "tuVf0wrAOVVlX/j0f6y0WOYAdjHba5P4qqTp1ZJDsQpmwZqm3BfiJz5DR6+aVQOY",
+            "Address": "t3w3sv7uykya4vkzk77d2h7lfuldtaa5rr3nvzh6fkutu5lesdwefgnqm2u3obpyrhhzbupl42kubzqmvfivjq"
+          },
+          "ProofType": 5
+        }
+      ]
+    }
+  ],
+  "NetworkName": "localnet-d39f7d65-d9c6-4e2e-8129-6f13dc096ea9",
+  "VerifregRootKey": {
+    "Type": "multisig",
+    "Balance": "0",
+    "Meta": {
+      "Signers": [
+        "t1ceb34gnsc6qk5dt6n7xg6ycwzasjhbxm3iylkiy"
+      ],
+      "Threshold": 1,
+      "VestingDuration": 0,
+      "VestingStart": 0
+    }
+  },
+  "RemainderAccount": {
+    "Type": "multisig",
+    "Balance": "0",
+    "Meta": {
+      "Signers": [
+        "t1ceb34gnsc6qk5dt6n7xg6ycwzasjhbxm3iylkiy"
+      ],
+      "Threshold": 1,
+      "VestingDuration": 0,
+      "VestingStart": 0
+    }
+  }
+}
+```
+
+### Generate a CAR file
+
+All that's left is for the `genesis.json` file to be turned into `genesis.car` by running the [genesis car](https://github.com/filecoin-project/lotus/blob/v1.17.0-rc3/cmd/lotus-seed/genesis.go#L559) command, which will create the Genesis block and the initial State Tree, and save it in the IPLD format that Forest expects to be started with.
+
+```bash
+docker run -it --rm \
+  -v $PWD/genesis-files:/out \
+  --user $(id -u):$(id -g) \
+  --entrypoint lotus-seed \
+  filecoin/lotus-test \
+    --sector-dir /out \
+    genesis car --out "/out/genesis.car" "/out/genesis.json"
+```
+
+The log helpfully tells us what is happening:
+
+```console
+2022-07-19T13:38:55.057Z	WARN	genesis	testing/genesis.go:57	Generating new random genesis block, note that this SHOULD NOT happen unless you are setting up new network
+init set t3w3sv7uykya4vkzk77d2h7lfuldtaa5rr3nvzh6fkutu5lesdwefgnqm2u3obpyrhhzbupl42kubzqmvfivjq t0100
+init set t1ceb34gnsc6qk5dt6n7xg6ycwzasjhbxm3iylkiy t0101
+[flexi_logger][ERRCODE::Time] flexi_logger has to work with UTC rather than with local time, caused by IndeterminateOffset
+    See https://docs.rs/flexi_logger/latest/flexi_logger/error_info/index.html#time
+publishing 1 storage deals on miner t01000 with worker t3w3sv7uykya4vkzk77d2h7lfuldtaa5rr3nvzh6fkutu5lesdwefgnqm2u3obpyrhhzbupl42kubzqmvfivjq
```

Comment:
How odd! The `init set ...` uses the worker key we saw in the JSON further up, but says `t0100` (ID=100), but a few lines further it says it's publishing 1 deal for miner `t01000` (ID=1000), which is what I expected to see.

Reply:
@adlrocha so are the miner ID and the actor ID separate things, do `t0100` and `t01000` both resolve to `t3w3...fivjq`, or just one of them? :dizzy_face:

Reply:
Correct. There's account actors and miner actors, and both can be controlled by the same keys. That is why we decided to just use account actors in Eudico. Every key address in the system has an on-chain account actor, and these keys can be control addresses for a miner actor. See https://github.com/adlrocha/builtin-actors/blob/9ede48182f541b565a63a294f3fc67a423128479/actors/miner/src/lib.rs#L142-L147

(I know, kind of a nightmare)

Reply:
Okay, so just to make sure I'm getting the whole picture.

As I understand we have `t01000` as the Miner ID, and `t0100` as the Account ID, right? 

* Can either of these be put in `BlockHeader.miner_address`? Sounds like it should be the Miner ID, but you are using the Account ID in eudico. 
* In Forest, The one in the header is passed to [get_miner_work_addr](https://github.com/ChainSafe/forest/blob/main/blockchain/state_manager/src/lib.rs#L264), which gets the `ActorState`, and treats it as `miner::State`, so it assumes it will be the Miner ID and not the Account ID. 
* `get_miner_work_addr` calls `resolve_to_key_addr` with the `worker`, which is the public key of the miner.
* `resolve_to_key_addr` will try to resolve _an address_ into a public key: if it's already the worker, which _is_ the public key, then it will just be returned (why does `get_miner_work_addr` even call it? can it be both in the state?), but if it's not a public key, it [will again load](https://github.com/ChainSafe/forest/blob/main/vm/interpreter/src/default_runtime.rs#L24-L30) the `ActorState` but this time treat it as `account::State`, which is expected to hold nothing else than - wait for it - an `Address` which the system somehow knows _must_ be a public key this time.

And this is what you mentioned at some point, that in Eudico you directly call `resolve_to_key_addr` and bypass `get_miner_work_addr` because it would try to use the wrong type to deserialise the actor state. In fact you, in eudico could even put the public key in the header as the `miner_address`, because `resolve_to_key_addr` would just return it, _but_ there would be no guarantee that an account state for it exists, _but_ there is nothing in the account state but an address anyway.

I might be missing something, but I don't see any difference in how `get_actor` resolves `t0100` vs `t01000`. Does that mean that lotus insists we start creating Miner IDs from 1000, but then assigns Account IDs for them from 100, and what, skips 1000 because it already exists when the 900th account is created? 

On a similar note, how can I resolve that the `t01000` Miner ID to the `t0100` Account ID?


Sorry for my eternal confusion on this topic, I appreciate your help trying to clear up the fog.

Reply:
It's definitely a bit of a mess. Let's see if I manage to clarify it here and if not we can try to have a quick sync.
- Every key address in the system has an underlying account actor (i.e. an on-chain identity, e.g. `t0100`). Miners, are controlled by one (or more) keys, and thus their corresponding account actors, and are identified on-chain through a miner actor (e.g. `t01000`). Thus, we have an account actor `t0100` that controls a miner actor `t01000`. 
- In the default implementation of Fiecoin (with a single consensus), blocks are signed with the keys of the worker ID of the miner, but we set in the `Miner` field of the block the on-chain identity of the miner (e.g. `t01000`). To validate the block, then, we first `get_miner_work_addr` to get the control address behind `t01000`, i.e. `t0100`, and we resolve the key address for the worker. These are the keys used to verify the block signature. So we go from miner actor ID, to ID of the account actor of the worker, to the public keys managing that account actor and used to sign the block.
- In Eudico, we don't need to have a miner actor to represent a validator (there is no storage involved). So instead of having a miner actor controlled by an account actor through a set of keys, we just have an account actor with validation rights. Thus, the `Miner` address includes the ID of an account actor, and to verify the signature we just need to `resolve_to_key_addr` (no control addresses or miner actors involved). 

Finally, IIRC the assignment of IDs in genesis works as follows: 
- `t00-t99` are reserved to system actors and special addresses (like the burn address).
- `t0100-t0999` are use to assign new account actors in genesis.
-  `t01000->onwards` are used to assign new actor throughout the life of the network (every new actor in the network after genesis is assigned the subsequent ID)

Let me know if it's clear now, if it's still not quite clear ping me for a quick sync (it may be more efficient).